docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,83 @@
+# Qualifying replicas for read operations
+
+## Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+the coordinator if some of the target replicas are in fact dead,
+overloaded to the point of unresponsiveness, having network problems, etc.
+
+## Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple failure detection
+mechanism with high granularity is implemented as follows:
+ * each time a read request is sent to a replica, this attempt's timestamp
+   is recorded
+ * first timestamp of a streak of unresponded requests is recorded separately
+   in order to see how long ago the coordinator started to try to communicate
+   with a replica
+ * each time a response is received from a replica, the round-trip time is recorded
+
+The combination of "last time a request was sent without response" and last successful
+round-trip is a cheap, rough implementation of a failure detector - if the node hasn't
+responded for more than X milliseconds since the previous request and the current request
+is going to time out in X milliseconds, the target node gets more and more likely to be
+qualified as dead.
```

Comment:
Again I think there's a confusion between "dead" and a node which because of some overload problem now has latency X for each request but it still very much alive. We should be clear which problem we're tackling.

Reply:
Yes, when I started writing this series and this document, I had both failure detection and dynamic snitch in mind, but then the scope got reduced to just failure detection. So ultimately I want to be able to detect nodes which are either dead or look dead (e.g. due to long network partition).

Reply:
And my intention with tying the detection with timeouts is to make the failure detector more strict for requests with short timeouts. The idea is that when a workload is full of requests with 200ms timeouts, it makes sense to have a more strict failure detection in order to minimize the time wasted on sending requests to dead replicas. On the other hand, long-standing requests with multi-second timeouts can be more lenient and give more nodes a chance to respond.